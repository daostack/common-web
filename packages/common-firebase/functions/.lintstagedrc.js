module.exports = {
  '**/*.js?(x)': (filenames) => `eslint ${filenames.join(' ')}`
}
