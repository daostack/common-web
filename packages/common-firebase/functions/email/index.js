const requestToJoinSubmitted = require('./templates/requestToJoinSubmitted');
const adminCommonCreated = require('./templates/adminCommonCreated');
const adminFundingRequestAccepted = require('./templates/adminFundingRequestAccepted');
const adminPreauthorizationFailed = require('./templates/adminPreauthorizationFailed');
const userCommonCreated = require('./templates/userCommonCreated');
const userCommonFeatured = require('./templates/userCommonFeatured');
const userFundingRequestAccepted = require('./templates/userFundingRequestAccepted');
const userJoinedButFailedPayment = require('./templates/userJoinedButFailedPayment');
const userJoinedSuccess = require('./templates/userJoinedSuccess');
const adminWalletCreationFailed = require('./templates/adminWalletCreationFailed');

const mailer = require("../mailer");

const templates = {
  requestToJoinSubmitted,
  adminCommonCreated,
  adminFundingRequestAccepted,
  adminPreauthorizationFailed,
  adminWalletCreationFailed,
  userCommonCreated,
  userCommonFeatured,
  userFundingRequestAccepted,
  userJoinedButFailedPayment,
  userJoinedSuccess
}

const globalDefaultStubs = {
  supportChatLink: 'https://common.io/help'
};

const replaceAll = (string, search, replace) => {
  return string.split(search).join(replace);
}

const isNullOrUndefined = (val) =>
  val === null || val === undefined;


/***
 *
 * @param {
 *  'requestToJoinSubmitted',
 *  'adminCommonCreated',
 *  'adminFundingRequestAccepted',
 *  'adminPreauthorizationFailed' ,
 *  'userCommonCreated',
 *  'userCommonFeatured',
 *  'userFundingRequestAccepted',
 *  'userJoinedButFailedPayment',
 *  'userJoinedSuccess',
 *  'adminWalletCreationFailed'
 * } templateKey
 *
 * @param {{
 *   emailStubs: object
 *   subjectStubs: object
 * }} payload
 *
 * @return {{
 *   template: string The templated body
 *   subject: string The templated subject
 * }}
 */
const getTemplatedEmail = (templateKey, payload) => {
  let {template, subject, emailStubs, subjectStubs} = templates[templateKey];

  if (isNullOrUndefined(template)) {
    throw new Error(`The requested template (${templateKey}) cannot be found`);
  }

  // Validate and add default values for the email template
  for (const stub in emailStubs || {}) {
    // Check if the stub is required. If it is check is there is value either in the
    // global stubs, the default stubs or the user provided ones
    if (
      emailStubs[stub].required && (
        isNullOrUndefined(payload.emailStubs[stub]) &&
        isNullOrUndefined(emailStubs[stub].default) &&
        isNullOrUndefined(globalDefaultStubs[stub])
      )
    ) {
      throw new Error(`Required stub ${stub} was not provided for email template`)
    }

    // If there is a default value for the stub and has not been replaced add it here
    if (!isNullOrUndefined(emailStubs[stub].default) && isNullOrUndefined(payload.emailStubs[stub])) {
      template = template.replace(`{{${stub}}}`, emailStubs[stub])
    } else if (!isNullOrUndefined(globalDefaultStubs[stub]) && isNullOrUndefined(payload.emailStubs[stub])) {
      template = template.replace(`{{${stub}}}`, globalDefaultStubs[stub])
    }
  }

  // Replace all provided stubs in the template
  for (const stub in payload.emailStubs) {
    template = replaceAll(template, `{{${stub}}}`, payload.emailStubs[stub])
  }

  // Validate the email subject
  for (const stub in subjectStubs) {
    if (subjectStubs[stub].required && isNullOrUndefined(payload.subjectStubs[stub])) {
      throw new Error(`Required stub ${stub} was not provided for subject template`)
    }
  }

  for (const stub in payload.subjectStubs) {
    subject = replaceAll(subject, `{{${stub}}}`, payload.subjectStubs[stub])
  }

  return {
    template,
    subject
  };
}

/***
 *
 * @param {
 *  'requestToJoinSubmitted',
 *  'adminCommonCreated',
 *  'adminFundingRequestAccepted',
 *  'adminPreauthorizationFailed' ,
 *  'userCommonCreated',
 *  'userCommonFeatured',
 *  'userFundingRequestAccepted',
 *  'userJoinedButFailedPayment',
 *  'userJoinedSuccess',
 *  'adminWalletCreationFailed'
 * } templateKey
 *
 * @param { object } emailStubs
 * @param { object } subjectStubs
 *
 * @param { string } to
 */
const sendTemplatedEmail = async ({ templateKey, emailStubs, subjectStubs, to}) => {
  const {template, subject} = getTemplatedEmail(templateKey, { emailStubs, subjectStubs });

  await mailer.sendMail(
    to,
    subject,
    template
  );
}

module.exports = {
  getTemplatedEmail,
  sendTemplatedEmail
}
