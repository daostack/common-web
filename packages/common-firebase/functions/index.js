require('module-alias/register');

const admin = require('firebase-admin');
const { databaseURL } = require('./settings');
const {env} = require('@env');

if(env.environment === 'dev') {
  admin.initializeApp();
} else {
  admin.initializeApp({
    credential: admin.credential.cert(require('@env/adminsdk-keys.json')),
    databaseURL: databaseURL
  });
}

const relayer = require('./relayer');
const graphql = require('./graphql');
const graphqlTriggers = require('./graphql/util/triggers');
const mangopay = require('./mangopay');
const mangopayTriggers = require('./mangopay/triggers');

// Add the __tests__ endpoints only if enabled
if(env.tests.enabled) {
  exports.tests = require('./tests').tests;
}

exports.relayer = relayer.relayer;
exports.graphql = graphql.graphql;
exports.mangopay = mangopay.mangopay;
exports.mangopaySubs = mangopayTriggers;
exports.graphqlSubs = graphqlTriggers;
// Disable notification
// exports.notification = require('./notification');

exports.cronJobs = require('./cron').crons;
