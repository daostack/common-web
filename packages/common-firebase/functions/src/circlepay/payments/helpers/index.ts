export { isFinalized, isSuccessful } from './statusHelper';

export { updatePayments } from './converter';

export { feesHelper } from './feesHelper';
export { failureHelper } from './failureHelper';