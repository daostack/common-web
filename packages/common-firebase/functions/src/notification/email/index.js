const requestToJoinSubmitted = require('./templates/requestToJoinSubmitted');
const adminCommonCreated = require('./templates/adminCommonCreated');
const adminFundingRequestAccepted = require('./templates/adminFundingRequestAccepted');
const adminPreauthorizationFailed = require('./templates/adminPreauthorizationFailed');
const userCommonCreated = require('./templates/userCommonCreated');
const userCommonFeatured = require('./templates/userCommonFeatured');
const userFundingRequestAccepted = require('./templates/userFundingRequestAccepted');
const userJoinedButFailedPayment = require('./templates/userJoinedButFailedPayment');
const userJoinedSuccess = require('./templates/userJoinedSuccess');
const adminWalletCreationFailed = require('./templates/adminWalletCreationFailed');
const adminJoinedButPaymentFailed = require('./templates/adminJoinedButFailedPayment');
const adminPayInSuccess = require('./templates/adminPayInSuccess');

const { CommonError } = require('../../util/errors');

const mailer = require('../../mailer');
const env = require('../../env').env;


const templates = {
  requestToJoinSubmitted,
  adminCommonCreated,
  adminFundingRequestAccepted,
  adminPreauthorizationFailed,
  adminWalletCreationFailed,
  userCommonCreated,
  userCommonFeatured,
  userFundingRequestAccepted,
  userJoinedButFailedPayment,
  userJoinedSuccess,
  adminJoinedButPaymentFailed,
  adminPayInSuccess
};

const globalDefaultStubs = {
  supportChatLink: 'https://common.io/help'
};

const replaceAll = (string, search, replace) => {
  return string.split(search).join(replace);
};

const isNullOrUndefined = (val) =>
  val === null || val === undefined;


/***
 *
 * @param {
 *  'requestToJoinSubmitted',
 *  'adminCommonCreated',
 *  'adminFundingRequestAccepted',
 *  'adminPreauthorizationFailed' ,
 *  'userCommonCreated',
 *  'userCommonFeatured',
 *  'userFundingRequestAccepted',
 *  'userJoinedButFailedPayment',
 *  'userJoinedSuccess',
 *  'adminWalletCreationFailed',
 *  'adminJoinedButPaymentFailed',
 *  'adminPayInSuccess'
 * } templateKey
 *
 * @param {{
 *   emailStubs: object
 *   subjectStubs?: object
 * }} payload
 *
 * @return {{
 *   template: string The templated body
 *   subject: string The templated subject
 * }}
 */
const getTemplatedEmail = (templateKey, payload) => {
  let { template, subject, emailStubs, subjectStubs } = templates[templateKey];

  console.debug('Email templating started');

  if (isNullOrUndefined(template)) {
    throw new CommonError(`The requested template (${templateKey}) cannot be found`);
  }

  // Validate and add default values for the email template
  for (const stub in emailStubs || {}) {
    // Check if the stub is required. If it is check is there is value either in the
    // global stubs, the default stubs or the user provided ones
    if (
      emailStubs[stub].required && (
        isNullOrUndefined(payload.emailStubs[stub]) &&
        isNullOrUndefined(emailStubs[stub].default) &&
        isNullOrUndefined(globalDefaultStubs[stub])
      )
    ) {
      throw new CommonError(`Required stub ${stub} was not provided for ${templateKey} template`);
    }

    // If there is a default value for the stub and has not been replaced add it here
    if (!isNullOrUndefined(emailStubs[stub].default) && isNullOrUndefined(payload.emailStubs[stub])) {
      template = template.replace(`{{${stub}}}`, emailStubs[stub]);
    } else if (!isNullOrUndefined(globalDefaultStubs[stub]) && isNullOrUndefined(payload.emailStubs[stub])) {
      template = template.replace(`{{${stub}}}`, globalDefaultStubs[stub]);
    }
  }

  // Replace all provided stubs in the template
  for (const stub in payload.emailStubs) {
    template = replaceAll(template, `{{${stub}}}`, payload.emailStubs[stub]);
  }

  // Validate the email subject
  for (const stub in subjectStubs) {
    if (subjectStubs[stub].required && isNullOrUndefined(payload.subjectStubs[stub])) {
      throw new CommonError(`Required stub ${stub} was not provided for subject template`);
    }
  }

  for (const stub in payload.subjectStubs) {
    subject = replaceAll(subject, `{{${stub}}}`, payload.subjectStubs[stub]);
  }

  console.debug(`Email templating finished for template ${templateKey}`);

  return {
    template,
    subject
  };
};

/***
 *
 * @param {
 *  'requestToJoinSubmitted',
 *  'adminCommonCreated',
 *  'adminFundingRequestAccepted',
 *  'adminPreauthorizationFailed' ,
 *  'adminJoinedButPaymentFailed',
 *  'userCommonCreated',
 *  'userCommonFeatured',
 *  'userFundingRequestAccepted',
 *  'userJoinedButFailedPayment',
 *  'userJoinedSuccess',
 *  'adminWalletCreationFailed',
 *  'adminPayInSuccess'
 * } templateKey
 *
 * @param { object } emailStubs
 * @param { object } subjectStubs
 *
 * @param { string | string[] | 'admin' } to
 */
const sendTemplatedEmail = async ({ templateKey, emailStubs, subjectStubs, to }) => {
  to === 'admin' && (to = env.mail.adminMail);

  const { template, subject } = getTemplatedEmail(templateKey, { emailStubs, subjectStubs });

  if (Array.isArray(to)) {
    const emailPromises = [];

    to.forEach((emailTo) => {
      console.log(`Sending ${templateKey} to ${emailTo}.`)

      emailPromises.push(mailer.sendMail({
        to: emailTo,
        subject,
        template
      }));
    });

    await Promise.all(emailPromises);
  } else {
    console.log(`Sending ${templateKey} to ${to}.`)

    await mailer.sendMail(
      to,
      subject,
      template
    );
  }


  console.log('Templated email send successfully');
};

module.exports = {
  getTemplatedEmail,
  sendTemplatedEmail,
};
