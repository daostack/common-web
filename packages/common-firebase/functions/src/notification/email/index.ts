import { sendMail } from './mailer';
import { env } from '../../constants';
import { CommonError } from '../../util/errors';

import { approvePayout } from './templates/approvePayout';
import { userCommonCreated } from './templates/userCommonCreated';
import { userJoinedSuccess } from './templates/userJoinedSuccess';
import { adminPayInSuccess } from './templates/adminPayInSuccess';
import { adminCommonCreated } from './templates/adminCommonCreated';
import { userCommonFeatured } from './templates/userCommonFeatured';
import { requestToJoinSubmitted } from './templates/requestToJoinSubmitted';
import { userFundingRequestAccepted } from './templates/userFundingRequestAccepted';
import { userJoinedButFailedPayment } from './templates/userJoinedButFailedPayment';
import { adminFundingRequestAccepted } from './templates/adminFundingRequestAccepted';
import { adminPreauthorizationFailed } from './templates/adminPreauthorizationFailed';
import { adminJoinedButPaymentFailed } from './templates/adminJoinedButFailedPayment';
import { subscriptionCanceled } from './templates/subscriptionCanceled';
import { subscriptionChargeFailed } from './templates/subscriptionChargeFailed';
import { subscriptionCharged } from './templates/subscriptionCharged';

const templates = {
  requestToJoinSubmitted,
  adminCommonCreated,
  adminFundingRequestAccepted,
  adminPreauthorizationFailed,
  userCommonCreated,
  userCommonFeatured,
  userFundingRequestAccepted,
  userJoinedButFailedPayment,
  userJoinedSuccess,
  adminJoinedButPaymentFailed,
  adminPayInSuccess,
  approvePayout,
  subscriptionCanceled,
  subscriptionCharged,
  subscriptionChargeFailed
};

const globalDefaultStubs = {
  supportChatLink: 'https://common.io/help'
};

const replaceAll = (string, search, replace) => {
  return string.split(search).join(replace);
};

const isNullOrUndefined = (val) =>
  val === null || val === undefined;

interface IStub {
  required: boolean;
  default?: string;
}

export interface IEmailTemplate {
  template: string;
  subject: string;

  emailStubs?: {
    [key: string]: IStub;
  };

  subjectStubs?: {
    [key: string]: IStub;
  };
}

interface ITemplatedEmail {
  body: string;
  subject: string;
}

// @todo Make the payload type based on the templateKey
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getTemplatedEmail = (templateKey: keyof typeof templates, payload: any): ITemplatedEmail => {
  let { template, subject } = templates[templateKey];
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  const { emailStubs, subjectStubs } = templates[templateKey];


  // @todo Logger is not definded here because the file is JS. Move it to TS
  // eslint-disable-next-line no-console
  console.debug('Email templating started');

  if (isNullOrUndefined(template)) {
    throw new CommonError(`The requested template (${templateKey}) cannot be found`);
  }

  // Validate and add default values for the email template
  for (const stub in emailStubs || {}) {
    // Check if the stub is required. If it is check is there is value either in the
    // global stubs, the default stubs or the user provided ones
    if (
      emailStubs[stub].required && (
        isNullOrUndefined(payload.emailStubs[stub]) &&
        isNullOrUndefined(emailStubs[stub].default) &&
        isNullOrUndefined(globalDefaultStubs[stub])
      )
    ) {
      throw new CommonError(`Required stub ${stub} was not provided for ${templateKey} template`);
    }

    // If there is a default value for the stub and has not been replaced add it here
    if (!isNullOrUndefined(emailStubs[stub].default) && isNullOrUndefined(payload.emailStubs[stub])) {
      template = template.replace(`{{${stub}}}`, emailStubs[stub]);
    } else if (!isNullOrUndefined(globalDefaultStubs[stub]) && isNullOrUndefined(payload.emailStubs[stub])) {
      template = template.replace(`{{${stub}}}`, globalDefaultStubs[stub]);
    }
  }

  // Replace all provided stubs in the template
  for (const stub in payload.emailStubs) {
    template = replaceAll(template, `{{${stub}}}`, payload.emailStubs[stub]);
  }

  // Validate the email subject
  for (const stub in subjectStubs) {
    if (subjectStubs[stub].required && isNullOrUndefined(payload.subjectStubs[stub])) {
      throw new CommonError(`Required stub ${stub} was not provided for subject template`);
    }
  }

  for (const stub in payload.subjectStubs) {
    subject = replaceAll(subject, `{{${stub}}}`, payload.subjectStubs[stub]);
  }

  // eslint-disable-next-line no-console
  console.debug(`Email templating finished for template ${templateKey}`);

  return {
    body: template,
    subject
  };
};

export interface ISendTemplatedEmailData {
  templateKey: keyof typeof templates,
  emailStubs?: any,
  subjectStubs?: any,
  to: string | string[]
}

type SendTemplatedEmail = (data: ISendTemplatedEmailData) => Promise<void>;

export const sendTemplatedEmail: SendTemplatedEmail = async ({ templateKey, emailStubs, subjectStubs, to }) => {
  to === 'admin' && (to = env.mail.adminMail);

  const { body, subject } = getTemplatedEmail(templateKey, { emailStubs, subjectStubs });

  if (Array.isArray(to)) {
    const emailPromises = [];

    to.forEach((emailTo) => {
      // eslint-disable-next-line no-console
      console.log(`Sending ${templateKey} to ${emailTo}.`);

      emailPromises.push(sendMail(
        emailTo,
        subject,
        body
      ));
    });

    await Promise.all(emailPromises);
  } else {
    // eslint-disable-next-line no-console
    console.log(`Sending ${templateKey} to ${to}.`);

    await sendMail(
      to,
      subject,
      body
    );
  }


  // eslint-disable-next-line no-console
  console.log('Templated email send successfully');
};

export default {
  getTemplatedEmail,
  sendTemplatedEmail
};
