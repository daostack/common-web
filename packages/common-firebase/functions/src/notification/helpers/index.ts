export const getFundingRequestAcceptedTemplate = (country: string, amount: number) => {
  if (amount) {
    return !country
      ? 'userFundingRequestAcceptedUnknown'
      : (country === 'IL'
        ? 'userFundingRequestAcceptedIsraeli'
        : 'userFundingRequestAcceptedForeign');
  }
  return 'userFundingRequestAcceptedZeroAmount';
};