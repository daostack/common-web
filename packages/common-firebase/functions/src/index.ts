import * as cron from './cron';
import * as event from './event';
import * as notification from './notification';
import * as circlepayTriggers from './circlepay/triggers';

import { circlepayApp } from './circlepay';
import { commonsApp } from './common';
import { proposalCrons, proposalsApp } from './proposals';

// --- Express apps
exports.commons = commonsApp;
exports.proposals = proposalsApp;
exports.circlepay = circlepayApp;

// --- Triggers and Subscriptions
exports.eventSub = event;
exports.notificationSub = notification;
exports.circlepayTriggers = circlepayTriggers;
exports.proposalCrons = proposalCrons;

exports.cronJobs = cron;
