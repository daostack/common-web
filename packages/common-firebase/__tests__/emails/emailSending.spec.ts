import { getTemplatedEmail } from '../../functions/src/email';
import { runTest } from '../helpers/runTest';
import '../../functions/src';

const requestToJoinStubs = {
  name: 'Test Name',
  link: 'https://google.com',
  commonName: 'Test Common Name'
};

const requestToJoinPartialStubs = {
  name: 'Test Name'
};

runTest(() => {
  it('should be successful with all stubs', () => {
    const templatedEmail = getTemplatedEmail('requestToJoinSubmitted', {
      emailStubs: requestToJoinStubs
    });

    expect(templatedEmail).not.toBe(null);
    expect(templatedEmail).not.toBe(undefined);
  });

  it('should produce the same email for the same stubs', () => {
    const templatedEmail = getTemplatedEmail('requestToJoinSubmitted', {
      emailStubs: requestToJoinStubs
    });

    expect(templatedEmail).toMatchSnapshot();
  });

  it('should throw on missing stub', () => {
    expect(() => {
      getTemplatedEmail('requestToJoinSubmitted', {
        emailStubs: requestToJoinPartialStubs
      })
    }).toThrow();
  });
})