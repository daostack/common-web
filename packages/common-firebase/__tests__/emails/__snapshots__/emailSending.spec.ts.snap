// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Functions should produce the same email for the same stubs 1`] = `
Object {
  "body": "
  Hello Test Name,<br />
  Thank you for Joining Common!
  <br /><br />
  Your request to join <a href=\\"https://google.com\\">Test Common Name</a> has been submitted and is pending approval by the Common members. 
  You will be notified by email once voting has concluded. 
  <br /><br />
  For more information you can contact us any time using our <a href=\\"https://common.io/help\\">support chat</a>
  <br />
  Common,<br />
  Collaborative Social Action.
",
  "subject": "Request to join submitted",
}
`;
